### DISCUSSION TITLE
Moose plastic hardening simulation - Computationaly slower than ANSYS

### DISCUSSION THREAD

**Discussion by abarun22** [idaholab/moose#16372] [Q&A General] [answered]
Dear all,
I recently used MOOSE  to predict the hardening behaviour of certain class of steel specimens subjected to tensile loads. A PieceWise linear function was used to model the hardening response. The results were matching precisely with ANSYS until the start of necking, with some differences noticed thereafter, nevertheless a seemingly realistic behaviour overall. However the problem is that MOOSE is extremely slow in comparison to ANSYS's computational effort.  For a model with say ~1500 nodes, Ansys took only 5-6 mins, whereas MOOSE clocks it in 2.5hrs, which is exhorbitant, considering that we may need to go for a constrained optimization problem thereafter. We typically expect this problem to be solved in 10 iterations and that might take more than a day to complete the whole analysis. From the compiler perspective, we've chosen the right optimization (O3) that enables a faster computation. Attached please find the input file i was using to simulate this model. Any advice on how to achieve a computational time comparable to ANSYS would be greatly appreciated?
Kind regards,
[PieceWise_test.txt](https://github.com/idaholab/moose/files/5631092/PieceWise_test.txt)

Arun

**Comment by jiangwen84** [comment]
Could you quickly check whether ANSYS and MOOSE are solving the same number of DOFs? 
How many time steps does MOOSE and ANSYS take? 
For each time step, how many nonlinear iteration does MOOSE and ANSYS?
What solver does ANSYS use? How long does ANSYS and MOOSE to solve per iteration and time step?

**Comment by srinath-chakravarthy** [comment]
my 2c. In my experience, MOOSE is generally slower than either ANSYS or ABAQUS.  For non-AD solutions in plasticity i have found moose to be about 2 - 2.5 times slower with a LU preconditioner and newton solution. 

The relative convergence criteria are about 1e-3 in ANSYS and ABAQUS. Also double check the elements you are using. The defaults in commercial software, use hybrid elements with an extra pressure degree of freedom, that is not yet implemented in MOOSE. Also turn on volumetric locking to get better convergence and speed. 

In your input file, change the solution to NEWTON instead of PJFNK and see if it helps. 

Cheers
Srinath

**Reply by jiangwen84** [reply]
For nonlinear mechanics problem, we do not implement exact Jacobians, so Newton should take take more iterations than PJFNK to converge. In our experience, for non-AD, NEWTON is usually slower than PJFNK. I am curious to know your experience on the speed of NEWTON and PJFNK for non-AD.

**Comment by dschwen** [comment, marked as answer]
Your `IsotropicPlasticityStressUpdate` is set to run on the displaced mesh. I don't think that's right. You might try using AD versions of all mechanics models `ADIsotropicPlasticityStressUpdate`, `ADComputeMultipleInelasticStress`, `ADComputeIsotropicElasticityTensor`, and `use_asutomatic_differentiation = true` in the Master action. (you may need to recompile your moose with a larger AD vector size for 3D...)

But the execution time difference is too substantial to be explained away by that most likely. Without the mesh and function files we cannot run your input though. 

1500 nodes is a really tiny problem and should absolutely not take that long to run _at all_.

**Reply by abarun22** [reply]
Dear all,
Thanks for the suggestions and indeed they worked very well. On top of that i resized my ANSYS mesh (~5600 degrees of freedom) to match closely with my MOOSE mesh. Also i was not exactly specifying the number of cores in my parallel execution launcher earlier and that took me to the serial mode.  With all these modifications i am now getting comparables times.  Ansys took 26 mins to finish the computation, while moose did that in 33mins, which seems quite satisfying going in to subsequent phase of activities. For your reference i am attaching here the required files for running the computation. Is there any way that i could tune the execution options to achieve an enhanced performance.
Kind regards,
Arun
[PlasticityModelling_files.zip](https://github.com/idaholab/moose/files/5638613/PlasticityModelling_files.zip)

**Reply by abarun22** [reply]
Thanks for all your suggestions. With much hope i am now testing my plasticity model for large meshes say of the order of 50k Dof's. Moose crossed that in 32 hrs. I've not exactly quantified the speed in Ansys as i am trying to sort out certain issues associated with the batch job execution internaly. My speculation is that ANSYS scales much slower than MOOSE for bigger models, perhaps would take double the time attained by MOOSE. I shall give you a clear update in this as we progress with ANSYS testing. 
Another issue i noted recently was that the newer version of MOOSE seems to be much slower than the earlier version (dated May'2020) with the former one requiring 55mins to complete the execution for the smaller test case i mentioned intially. From my tests I see  that the petsc versions are not influencing the performance. I wonder whether the new developments in MOOSE are reponsible for this degraded performance. Any advise on this would help me decide on the performance aspects of the code before going in to the subsequent phase of developments.
Kind regards,
Arun

**Reply by hugary1995** [reply]
@abarun22 It's probably better to start a new discussion on the slow down using recent moose, so that people know this is _unanswered_.
